<a href="https://colab.research.google.com/github/NahinAlam001/NahinApp/blob/main/Book_Recommender_System.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1>Mounting to drive</h1>

In [74]:
#Connecting to drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


<h1>Importing Libraries<h1>

In [75]:
#Importing some common libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

<h1>Reading the dataset and preprocessing </h1>


In [76]:
#Reading the books dataset
books = pd.read_csv('/content/drive/MyDrive/Book_Recommender_System/Dataset/Books.csv',  on_bad_lines='skip', encoding='latin-1')

<ipython-input-76-e043d0e1dd43>:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv('/content/drive/MyDrive/Book_Recommender_System/Dataset/Books.csv',  on_bad_lines='skip', encoding='latin-1')


In [77]:
books.head(5)

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [78]:
books.shape

(271360, 8)

In [79]:
users = pd.read_csv('/content/drive/MyDrive/Book_Recommender_System/Dataset/Users.csv', on_bad_lines='skip', encoding='latin-1')

In [80]:
users.head(5)

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [81]:
users.shape

(278858, 3)

In [82]:
ratings = pd.read_csv('/content/drive/MyDrive/Book_Recommender_System/Dataset/Ratings.csv', on_bad_lines='skip', encoding='latin-1')

In [83]:
ratings.head(5)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [84]:
ratings.shape

(1149780, 3)

In [85]:
#Counting the number of ratings given by each user
ratings['User-ID'].value_counts()

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
116180        1
116166        1
116154        1
116137        1
276723        1
Name: User-ID, Length: 105283, dtype: int64

<h3>Here we can see that some of the users have only provided 1 rating. We can not cluster such users as 1 rating is not sufficient to find out about their taste. Those users who have rated less than 100 books should be filtered out since their information will not be very effective.</h3>

In [86]:
#Number of users who have rated more than 100 books
effective_users = ratings['User-ID'].value_counts() > 100

In [87]:
effective_users[effective_users].shape

(1825,)

In [88]:
y = effective_users[effective_users].index

In [89]:
y

Int64Index([ 11676, 198711, 153662,  98391,  35859, 212898, 278418,  76352,
            110973, 235105,
            ...
            238186,  99441, 187410, 262070,  70183,  40553,  39345, 266283,
            189666, 140879],
           dtype='int64', length=1825)

In [90]:
ratings = ratings[ratings['User-ID'].isin(y)]

In [91]:
ratings.shape

(656605, 3)

In [92]:
BookwithRatings = ratings.merge(books, on="ISBN")

In [93]:
BookwithRatings.head(5)

User-ID        ISBN  Book-Rating  \
0   276925  002542730X           10   
1   277427  002542730X           10   
2     3363  002542730X            0   
3    10030  002542730X            7   
4    11676  002542730X            6   

                                          Book-Title        Book-Author  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   

  Year-Of-Publication                  Publisher  \
0                1994  John Wiley &amp; Sons Inc   
1                1994  John Wiley &amp; Sons Inc   
2                1994  John Wiley &amp; Sons Inc   
3                1994  John Wiley &amp; Sons Inc   
4                1994  John Wiley &amp; Sons Inc   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/002542730X.0...   
1  http://images.amazon.com/images/P/002542730X.0...   
2  http://images.amazon.com/images/P/002542730X.0...   
3  http://images.amazon.com/images/P/002542730X.0...   
4  http://images.amazon.com/images/P/002542730X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/002542730X.0...   
1  http://images.amazon.com/images/P/002542730X.0...   
2  http://images.amazon.com/images/P/002542730X.0...   
3  http://images.amazon.com/images/P/002542730X.0...   
4  http://images.amazon.com/images/P/002542730X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/002542730X.0...  
1  http://images.amazon.com/images/P/002542730X.0...  
2  http://images.amazon.com/images/P/002542730X.0...  
3  http://images.amazon.com/images/P/002542730X.0...  
4  http://images.amazon.com/images/P/002542730X.0...

In [94]:
#Dropping uncessary columns
BookwithRatings=BookwithRatings.drop(['Image-URL-S','Image-URL-M', 'Image-URL-L'], axis=1)
BookwithRatings.head(5)

User-ID        ISBN  Book-Rating  \
0   276925  002542730X           10   
1   277427  002542730X           10   
2     3363  002542730X            0   
3    10030  002542730X            7   
4    11676  002542730X            6   

                                          Book-Title        Book-Author  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   

  Year-Of-Publication                  Publisher  
0                1994  John Wiley &amp; Sons Inc  
1                1994  John Wiley &amp; Sons Inc  
2                1994  John Wiley &amp; Sons Inc  
3                1994  John Wiley &amp; Sons Inc  
4                1994  John Wiley &amp; Sons Inc

In [95]:
BookwithRatings.shape

(604854, 7)

<h3>We got 604854 data after merging rating with book information.</h3>

In [96]:
num_rating = BookwithRatings.groupby('Book-Title')['Book-Rating'].count().reset_index()

In [97]:
num_rating.rename(columns = {'Book-Rating' : 'No_of_ratings'}, inplace = True)

In [98]:
num_rating.head()

Book-Title  No_of_ratings
0   A Light in the Storm: The Civil War Diary of ...              3
1                              Always Have Popsicles              1
2               Apple Magic (The Collector's series)              1
3   Beyond IBM: Leadership Marketing and Finance ...              1
4   Clifford Visita El Hospital (Clifford El Gran...              1

<h3>We can see that some of the books got only 1-3 ratings. These type of data needs to be removed as well otherwise there will be imbalance. We will keep only those books which has atleast 50 ratings.</h3>

In [99]:
BookwithRatings = BookwithRatings.merge(num_rating, on='Book-Title')

In [100]:
BookwithRatings = BookwithRatings[BookwithRatings['No_of_ratings']>=50]

In [101]:
BookwithRatings.sample(10)

User-ID        ISBN  Book-Rating  \
193210    35859  0451151259            0   
193771    53174  0451190572            0   
14738     93363  0553280341            0   
103589   143415  0440224837            0   
131349   145619  0312980159            0   
3199     224349  0140067477            0   
264952   267409  155166674X            0   
92335    249894  1592400876            8   
71973    239594  0345387651            7   
127638    76352  0452283205            0   

                                               Book-Title      Book-Author  \
193210                                 Eyes of the Dragon     Stephen King   
193771  The Green Mile: Coffey on the Mile (Green Mile...     Stephen King   
14738   B Is for Burglar (Kinsey Millhone Mysteries (P...      Sue Grafton   
103589                                 The Long Road Home   Danielle Steel   
131349                                         Fast Women  Jennifer Crusie   
3199                                      The Tao of Pooh    Benjamin Hoff   
264952                                    Between Friends  Debbie Macomber   
92335   Eats, Shoots &amp; Leaves: The Zero Tolerance ...      Lynne Truss   
71973                               The Cider House Rules      John Irving   
127638                                     Falling Angels  Tracy Chevalier   

       Year-Of-Publication                Publisher  No_of_ratings  
193210                1988      Penguin Putnam~mass             50  
193771                1996              Signet Book             51  
14738                 1986                   Bantam            166  
103589                1999  Dell Publishing Company             74  
131349                2002       St. Martin's Press            106  
3199                  1983            Penguin Books            100  
264952                2003               Mira Books             61  
92335                 2004             Gotham Books             69  
71973                 1994         Ballantine Books            197  
127638                2002              Plume Books             83

In [102]:
BookwithRatings.shape

(100426, 8)

In [103]:
BookwithRatings.drop_duplicates(['User-ID','Book-Title'], inplace=True)

In [104]:
BookwithRatings.shape

(97962, 8)

<h3>Finally we have 97962 data</h3>

In [105]:
pivot_table = BookwithRatings.pivot_table(columns='User-ID', index = 'Book-Title', values = 'Book-Rating')

In [106]:
pivot_table.shape

(1094, 1794)

In [107]:
pivot_table.fillna(0, inplace=True)

In [108]:
pivot_table

User-ID                                             183     254     507     \
Book-Title                                                                   
1984                                                   0.0     9.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2010: Odyssey Two                                      0.0     0.0     0.0   
204 Rosewood Lane                                      0.0     0.0     0.0   
24 Hours                                               0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             882     1424    1435    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2010: Odyssey Two                                      0.0     0.0     0.0   
204 Rosewood Lane                                      0.0     0.0     0.0   
24 Hours                                               0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     7.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             1733    1903    2033    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2010: Odyssey Two                                      0.0     0.0     0.0   
204 Rosewood Lane                                      0.0     0.0     0.0   
24 Hours                                               0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2110    ...  276018  \
Book-Title                                                  ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2010: Odyssey Two                                      0.0  ...     0.0   
204 Rosewood Lane                                      0.0  ...     0.0   
24 Hours                                               0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [109]:
#WE CAN USE CSR MATRIX TO ENSURE 0 WON'T BE CONSIDERED

In [110]:
from scipy.sparse import csr_matrix

In [111]:
final_data = csr_matrix(pivot_table)

In [112]:
final_data

<1094x1794 sparse matrix of type '<class 'numpy.float64'>'
	with 26868 stored elements in Compressed Sparse Row format>

In [113]:
from sklearn.neighbors import NearestNeighbors
model = NearestNeighbors(algorithm = 'brute')

In [114]:
model.fit(final_data)

NearestNeighbors(algorithm='brute')

In [115]:
book_names = pivot_table.index

In [116]:
import pickle
pickle.dump(model, open('/content/drive/MyDrive/Book_Recommender_System/SavedModel/model.pkl', 'wb'))
pickle.dump(book_names, open('/content/drive/MyDrive/Book_Recommender_System/SavedModel/books_name.pkl', 'wb'))
pickle.dump(final_data, open('/content/drive/MyDrive/Book_Recommender_System/SavedModel/final_data.pkl', 'wb'))
pickle.dump(pivot_table, open('/content/drive/MyDrive/Book_Recommender_System/SavedModel/pivot_table.pkl', 'wb'))

In [117]:
def recommend_book(book_name):
  book_id = np.where(pivot_table.index == book_name)[0][0]
  distance, suggestion = model.kneighbors(pivot_table.iloc[book_id,:].values.reshape(1,-1), n_neighbors=5)
  for i in range(len(suggestion)):
    books = pivot_table.index[suggestion[i]]
    for j in books:
      print(j)

In [118]:
book_name = '1984'
recommend_book(book_name)

1984
Tom Clancy's Op-Center: Games of State (Tom Clancy's Op Center (Paperback))
Ground Zero and Beyond
Thousand Acres
Darkfall


In [124]:
from sklearn.cluster import KMeans

In [125]:
# Create a KMeans object with 5 clusters
kmeans = KMeans(n_clusters=5)

In [126]:
# Fit the KMeans object to the pivot table data
kmeans.fit(pivot_table)

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=5)

In [127]:
# Predict the cluster label for each book
book_cluster_labels = kmeans.predict(pivot_table)

In [133]:
# Identify the books in the same cluster as the book that the user is interested in
def recommend_book_kmeans(book_name):
  if book_name in pivot_table.index:
    book_id = np.where(pivot_table.index == book_name)[0][0]
    book_cluster_label = book_cluster_labels[book_id]

    # Find all books in the same cluster
    cluster_books = pivot_table.index[book_cluster_labels == book_cluster_label]

    # Recommend the books in the same cluster to the user
    for book in cluster_books:
      print(book)
  else:
    print(f"Book '{book_name}' not found in the dataset.")


In [136]:
# Get the book name that the user is interested in
book_name = input('Enter the book name that you are interested in: ')

Enter the book name that you are interested in: Darkfall


In [137]:
# Recommend books using K-Means clustering
recommend_book_kmeans(book_name)

1984
2010: Odyssey Two
204 Rosewood Lane
24 Hours
4 Blondes
84 Charing Cross Road
A 2nd Helping of Chicken Soup for the Soul (Chicken Soup for the Soul Series (Paper))
A Beautiful Mind: The Life of Mathematical Genius and Nobel Laureate John Nash
A Case of Need
A Civil Action
A Confederacy of Dunces (Evergreen Book)
A Dangerous Fortune
A Darkness More Than Night
A Day Late and a Dollar Short
A Fine Balance
A Girl Named Zippy: Growing Up Small in Mooreland Indiana (Today Show Book Club #3)
A Great Deliverance
A Heartbreaking Work of Staggering Genius
A Is for Alibi (Kinsey Millhone Mysteries (Paperback))
A Lesson Before Dying (Vintage Contemporaries (Paperback))
A Light in the Window (The Mitford Years)
A Man Named Dave: A Story of Triumph and Forgiveness
A Man in Full
A Map of the World
A Patchwork Planet
A Prayer for Owen Meany
A Separate Peace
A Simple Plan
A Thousand Acres (Ballantine Reader's Circle)
A Virtuous Woman (Oprah's Book Club (Paperback))
A Walk in the Woods: Rediscoverin